misc/metrics/src/identify.rs

```diff
@@ -213,55 +175,104 @@ struct AddressLabels {
     protocols: String,
 }
 
-#[derive(Default, Clone, Debug)]
-struct Protocols {
-    peers: Arc<Mutex<HashMap<PeerId, Vec<String>>>>,
-}
+#[derive(Default, Debug, Clone)]
+struct Peers(Arc<Mutex<HashMap<PeerId, libp2p_identify::Info>>>);
 
-impl Protocols {
-    fn add(&self, peer: PeerId, protocols: Vec<String>) {
-        self.peers
-            .lock()
-            .expect("Lock not to be poisoned")
-            .insert(peer, protocols);
+impl Peers {
+    fn record(&self, peer_id: PeerId, info: libp2p_identify::Info) {
+        self.0.lock().unwrap().insert(peer_id, info);
     }
 
-    fn remove(&self, peer: PeerId) {
-        self.peers
-            .lock()
-            .expect("Lock not to be poisoned")
-            .remove(&peer);
+    fn remove(&self, peer_id: PeerId) {
+        self.0.lock().unwrap().remove(&peer_id);
     }
 }
 
-impl EncodeMetric for Protocols {
-    fn encode(&self, mut encoder: MetricEncoder) -> Result<(), std::fmt::Error> {
-        let count_by_protocol = self
-            .peers
-            .lock()
-            .expect("Lock not to be poisoned")
-            .iter()
-            .fold(
-                HashMap::<String, i64>::default(),
-                |mut acc, (_, protocols)| {
-                    for protocol in protocols {
-                        let count = acc.entry(protocol.to_string()).or_default();
-                        *count += 1;
-                    }
-                    acc
-                },
-            );
+impl Collector for Peers {
+    fn collect<'a>(
+        &'a self,
+    ) -> Box<dyn Iterator<Item = (Cow<'a, Descriptor>, MaybeOwned<'a, Box<dyn LocalMetric>>)> + 'a>
+    {
+        let mut count_by_protocols: HashMap<String, i64> = Default::default();
+        let mut count_by_listen_addresses: HashMap<String, i64> = Default::default();
+        let mut count_by_observed_addresses: HashMap<String, i64> = Default::default();
+
+        for (_, peer_info) in self.0.lock().unwrap().iter() {
+            {
+                let mut protocols: Vec<_> = peer_info
+                    .protocols
+                    .iter()
+                    .map(|p| {
+                        if ALLOWED_PROTOCOLS.contains(&p.as_bytes()) {
+                            p.to_string()
+                        } else {
+                            "unrecognized".to_string()
+                        }
+                    })
+                    .collect();
+                protocols.sort();
+                protocols.dedup();
```

Comment:
Could we extract this into a helper and thus make the loop a bit easier to read?